libnavigation-core/src/main/java/com/mapbox/navigation/core/routealternatives/RouteAlternativesController.kt

```diff
@@ -0,0 +1,110 @@
+import kotlinx.coroutines.launch
+import java.util.concurrent.CopyOnWriteArraySet
+
+internal class RouteAlternativesController(
+    private val options: RouteAlternativesOptions,
+    private val navigator: MapboxNativeNavigator,
+    private val directionsSession: DirectionsSession,
+    private val tripSession: TripSession,
+    private val routeOptionsUpdater: RouteOptionsUpdater
+) {
+    private val jobControl = ThreadController.getMainScopeAndRootJob()
+
+    private val mapboxTimer = MapboxTimer()
+    private val observers = CopyOnWriteArraySet<RouteAlternativesObserver>()
+    private var currentRequestId: Long? = null
+
+    fun register(routeAlternativesObserver: RouteAlternativesObserver) {
+        val needsToStartTimer = observers.isEmpty()
+        observers.add(routeAlternativesObserver)
+        if (needsToStartTimer) {
+            mapboxTimer.restartAfterMillis = options.intervalMillis
+            mapboxTimer.startTimer {
+                requestRouteAlternatives()
+            }
+        }
+    }
+
+    fun unregister(routeAlternativesObserver: RouteAlternativesObserver) {
+        observers.remove(routeAlternativesObserver)
+        if (observers.isEmpty()) {
+            stop()
+        }
+    }
+
+    fun unregisterAll() {
+        observers.clear()
+        stop()
+    }
+
+    private fun stop() {
+        mapboxTimer.stopJobs()
+        jobControl.job.cancelChildren()
+        currentRequestId?.let { directionsSession.cancelRouteRequest(it) }
+    }
+
+    private fun requestRouteAlternatives() {
+        if (directionsSession.routes.isEmpty()) {
+            return
+        }
+
+        val routeOptionsResult = routeOptionsUpdater.update(
+            directionsSession.getPrimaryRouteOptions(),
+            tripSession.getRouteProgress(),
+            tripSession.getEnhancedLocation()
+        )
+
+        when (routeOptionsResult) {
+            is RouteOptionsUpdater.RouteOptionsResult.Success -> {
+                currentRequestId?.let { directionsSession.cancelRouteRequest(it) }
+                currentRequestId = directionsSession.requestRoutes(
+                    routeOptionsResult.routeOptions,
+                    routesRequestCallback
+                )
+            }
+            is RouteOptionsUpdater.RouteOptionsResult.Error -> {
+                logger.e(
+                    msg = Message("Route alternatives options are not available"),
+                    tr = routeOptionsResult.error
+                )
+            }
+        }
+    }
+
+    private val routesRequestCallback = object : RoutesRequestCallback {
+        override fun onRoutesReady(routes: List<DirectionsRoute>) {
+            val routeProgress = tripSession.getRouteProgress()
+                ?: return
+            jobControl.scope.launch {
+                val alternatives = routes.filter { navigator.isDifferentRoute(it) }
+                observers.forEach { it.onRouteAlternatives(routeProgress, alternatives) }
```

Comment:
Check if the trip session is active and if there is an active route (basically check if we're in active guidance, you can also expose `NavigationSession#state` internally) before calling the observers.

Reply:
yeah added this with the other suggestion https://github.com/mapbox/mapbox-navigation-android/pull/4375#discussion_r634329821